llama_index/indices/service_context.py

```diff
@@ -71,15 +76,31 @@ def from_defaults(
         If an argument is specified, then use the argument value provided for that
         parameter. If an argument is not specified, then use the default value.
 
+        You can change the base defaults by setting llama_index.global_service_context
+        to a ServiceContext object with your desired settings.
+
         Args:
             llm_predictor (Optional[BaseLLMPredictor]): LLMPredictor
             prompt_helper (Optional[PromptHelper]): PromptHelper
             embed_model (Optional[BaseEmbedding]): BaseEmbedding
             node_parser (Optional[NodeParser]): NodeParser
             llama_logger (Optional[LlamaLogger]): LlamaLogger (deprecated)
             chunk_size_limit (Optional[int]): chunk_size_limit
+            callback_manager (Optional[CallbackManager]): CallbackManager
 
         """
+        if llama_index.global_service_context is not None:
```

Comment:
answering my own question: looks like when global service context is specified, when we call `from_defaults`, we get a shallow copy of the service context, where the same instance of e.g. callback manager would be used.

This behavior is different from when `global_service_context` is not specified, since we are making a new copy of each thing?

Not sure if there's any practical implication. If it's possible to make the two behaviors more similar without much effort. Would prefer that, but otherwise this is fine.

Reply:
Yea so I think of this as a sort of hierarchy

The values you pass into `from_defaults()` get used. For attributes you do not set, we pull from the global service context. And for when you don't have a global service context, then we use the previous defaults.

from_defaults -> global_service_context -> current hardcoded defaults